src/device-api/v2.ts

```diff
@@ -22,97 +18,82 @@ import * as images from '../compose/images';
 import * as volumeManager from '../compose/volume-manager';
 import * as serviceManager from '../compose/service-manager';
 import { spawnJournalctl } from '../lib/journald';
-import {
-	appNotFoundMessage,
-	serviceNotFoundMessage,
-	v2ServiceEndpointInputErrorMessage,
-} from './messages';
 import log from '../lib/supervisor-console';
 import supervisorVersion = require('../lib/supervisor-version');
-import { checkInt, checkTruthy } from '../lib/validation';
+import { checkInt, checkString, checkTruthy } from '../lib/validation';
+import {
+	isNotFoundError,
+	isBadRequestError,
+	BadRequestError,
+} from '../lib/errors';
 import { isVPNActive } from '../network';
-import * as actions from './actions';
 import { AuthorizedRequest } from './api-keys';
 import { fromV2TargetState } from '../lib/legacy';
+import * as actions from './actions';
+import { v2ServiceEndpointError } from './messages';
 
 export const router = express.Router();
 
-const handleServiceAction = (
-	req: AuthorizedRequest,
-	res: Response,
-	next: NextFunction,
-	action: CompositionStepAction,
-): Resolvable<void> => {
-	const { imageId, serviceName, force } = req.body;
-	const appId = checkInt(req.params.appId);
-	if (!appId) {
-		res.status(400).json({
-			status: 'failed',
-			message: 'Missing app id',
-		});
-		return;
-	}
-
-	// handle the case where the appId is out of scope
-	if (!req.auth.isScoped({ apps: [appId] })) {
-		res.status(401).json({
-			status: 'failed',
-			message: 'Application is not available',
-		});
-		return;
-	}
+const handleServiceAction = (action: CompositionStepAction) => {
+	return async (req: AuthorizedRequest, res: Response, next: NextFunction) => {
+		const [appId, imageId, serviceName, force] = [
+			checkInt(req.params.appId),
+			checkInt(req.body.imageId),
+			checkString(req.body.serviceName),
+			checkTruthy(req.body.force),
+		];
 
-	return Promise.all([applicationManager.getCurrentApps(), getApp(appId)])
-		.then(([apps, targetApp]) => {
-			const app = apps[appId];
+		if (!appId) {
+			return res.status(400).json({
+				status: 'failed',
+				message: 'Invalid app id',
+			});
+		}
 
-			if (app == null) {
-				res.status(404).send(appNotFoundMessage);
-				return;
-			}
+		if (!req.auth.isScoped({ apps: [appId] })) {
+			return res.status(401).json({
+				status: 'failed',
+				message: 'Application is not available',
```

Comment:
I know these messages come from before us, but maybe `Unauthorized` is the right message here?

Reply:
Good call, changing just the message doesn't break the interface. I changed all instances of JSON 401's that were returning other messages to `Unauthorized`. There were 3 instances of this in v1.ts and 4 instances in v2.ts. See https://github.com/balena-os/balena-supervisor/pull/2048/commits/c6cf6a013610a2ec8eba8149ca17f0375324bc33